rc/tools/git.kak

```diff
@@ -71,22 +76,25 @@ define-command -params 1.. \
     if [ $kak_token_to_complete -eq 0 ]; then
         printf "add\nrm\nreset\nblame\ncommit\ncheckout\ndiff\nhide-blame\nhide-diff\nlog\nnext-hunk\nprev-hunk\nshow\nshow-branch\nshow-diff\ninit\nstatus\nupdate-diff\ngrep\n"
     else
+        # Try cd-ing into the git work tree if exists
+        cd "$kak_opt_git_work_tree_top_level" &> /dev/null
         case "$1" in
             commit) printf -- "--amend\n--no-edit\n--all\n--reset-author\n--fixup\n--squash\n"; git ls-files -m ;;
             add) git ls-files -dmo --exclude-standard ;;
             rm|grep) git ls-files -c ;;
         esac
     fi
   } \
-  git %{ evaluate-commands %sh{
-    cd_bufdir() {
-        dirname_buffer="${kak_buffile%/*}"
-        cd "${dirname_buffer}" 2>/dev/null || {
-            printf 'fail Unable to change the current working directory to: %s\n' "${dirname_buffer}"
-            exit 1
-        }
-    }
-
+  git %{
+  set-option global git_work_tree_top_level %sh{
+    # - try to find a toplevel for a given buffile
+    # - if failed, test the current toplevel variable and if it is a functional git repository
+    # - if failed, fallback to `pwd`
+    ( cd "${kak_buffile%/*}" && git rev-parse --show-toplevel ) 2>/dev/null || \
+    ( cd "${kak_opt_git_work_tree_top_level}" && git rev-parse --show-toplevel ) 2>/dev/null || \
```

Comment:
I sometimes do

	kak src/main.cc
	:edit ../some-other-git-worktree/file.c

but I still expect `:git diff` to run in cwd instead of `some-other-git-worktree`.

I'm not yet sure which use case is more important.
I appreciate your DWIM approach but I'd like to think about the UI a bit more.

One quick workaround for editing multiple repositories is to run (perhaps from a mapping or even a `WinDisplay` hook?): `cd %sh{git -C "${kak_buffile%/*}" rev-parse --show-toplevel}`.
Or would that disrupt your workflow?

If we implemented `setenv`, one could do

	setenv GIT_WORKTREE /path/to/some-git-worktree

which works for all git commands

Reply:
Thanks for a quick response :)

> but I still expect `:git diff` to run in cwd instead of `some-other-git-worktree`.

I was expecting that I will break someone's workflow :P

> One quick workaround for editing multiple repositories is to run (perhaps from a mapping or even a `WinDisplay` hook?): `cd %sh{git -C "${kak_buffile%/*}" rev-parse --show-toplevel}`.
Or would that disrupt your workflow?

I usually keep a long running kakoune session with LSP configured by the kakscript located at the workspace path that gets loaded automatically and which configures multiple projects at the same time
```
set global lsp_config %{
    [language.rust.settings.rust-analyzer]
    linkedProjects = [
        "<project-1>/Cargo.toml",
        "<project-2>/Cargo.toml",
    ]
# ...
}
```
plus I usually have an `.ignore` file there as well that gets picked up by ripgrep and fd-find.

`cd`-ing everytime I want to `:git ..` is probably not great as it will confuse LSP and restoring the previous workspace path is probably quite cumbersome. I like being able to `:fzf-find` or `:fzf-grep` from anywhere, I very rarely do `:e` and similar.

My idea assumes some level of statefulness so random fileless buffers, like *grep*, *git* and such will inherit/pick up the worktree that was used on the very last "normal" file. Maybe with the envvars being controllable it would be as easy as setting `GIT_DIR` and `GIT_WORK_TREE` in a hook with the same logic as in the `git_work_tree_top_level` setter now. At least it sounds like a viable path.

Reply:
@krobelus I think I restored previous behaviour with the latest commit. Now, if the `git` repository is reachable from within the `cwd`, it will have unconditional precedence. So your workflow is preserved. However, if there's no repository from `cwd`, we try to be more clever and look at the file. I left the `pwd` fallback to preserve errors coming from `git` the way they worked so far. Do you think it makes sense now? :)

Reply:
Interesting workflow. Is the linkedProjects settings equivalent to defining a cargo workspace?

I need to think about which behavior is better overall.
Right now I personally prefer  that `git status` and `git diff` run in cwd instead of the buffer's git worktree. 

Falling back to the buffer's git worktree seems benign either way. Of course it will fail if `$HOME` is a git worktree, which is unfortunate

I think Emacs with Projectile will behave like your initial patch (didn't check though).

I'll try both options. The behavior change wouldn't be a tragic loss. I think it also depends on what other rc scripts expect.
It is possible that they can benefit from a project root. Personally I haven't dared to use commands like :git for projects below cwd.

Reply:
ok I hit this recently; I think I'm coming around on this one.. I think running in the buffer's worktree is the better default behavior.
However if `$PWD` is a child directory of the buffer's worktree, we should not `cd` up because that could invalidate pathspecs and other things.

So I think the logic should be

```sh
if ! bufworktree=$(git -C "${kak_buffile%/*}" rev-parse --show-toplevel); then
	exit # use $PWD
fi
if [ "${PWD#"$bufworktree"}" != "$PWD" ]; then
	# $PWD is a child of a Git worktree
	exit # use $PWD
fi
cd "$bufworktree"
```

If we really want to add a state variable on top of that, I'd like to hear the motivation.
(we can add support for `:git -Cpath/to/worktree` and GIT_DIR/GIT_WORK_TREE env vars if needed)

If you're too busy, I'm happy to take over this PR, LMK

Reply:
Sorry, I disappeared for a longer while. All kinds of errands happening. I really appreciate the input. 

> If you're too busy, I'm happy to take over this PR, LMK

Sure, if you feel like playing with it, that'd be great. I'm not that great of a shell expert anyhow but I'd like to participate and give feedback.

> logic should be

> state variable

A lot of discussion happened across the comments but I understand the global option being evil part. I liked one aspect of my solution that maintained the "information about the latest reasonable git work tree path". Which could be used anywhere for any kind of scripting. It was especially nice for scratch buffers like *git* so one was not losing the information about "what was the relevant git work tree" a moment ago.